tools/wptrunner/wptrunner/executors/executorwebdriver.py

```diff
@@ -897,17 +897,28 @@ async def process_bidi_event(method, params):
             # Use protocol loop to run the async cleanup.
             protocol.loop.run_until_complete(protocol.bidi_events.unsubscribe_all())
 
-        extra = {}
-        if (leak_part := getattr(protocol, "leak", None)) and (counters := leak_part.check()):
```

Comment:
Is it possible we make LeakProtocolPart part of ChromeDriverProtocol and WebDriverProtocol, and the check is only turned on when the command line argument is present?

Reply:
The layering is awkward because `LeakProtocolPart` isn't backed by any standard WebDriver APIs (unlike all the other `ProtocolPart`s belonging to `WebDriverProtocol`). Ideally, `LeakProtocolPart` wouldn't even be mentioned in this file, and `chrome` would do something similar to [`MarionetteAssertsProtocolPart`][0] where the calling code lives entirely in `executorchrome.py`. But the hooks that exist in `WebDriver*Executor` currently aren't sufficient to do this, short of forking the entire class.

[0]: https://github.com/web-platform-tests/wpt/blob/88831ee6202f4f23bc0e5601fd359de2ffe063f2/tools/wptrunner/wptrunner/executors/executormarionette.py#L355

Reply:
I think the goal is to make the architecture cleaner and easier to extend, for example what if Firefox want to do similar things. But as this works now, this can land the change for now.

I would hope somebody can take a effect to standardize this.

Reply:
> I think the goal is to make the architecture cleaner and easier to extend, for example what if Firefox want to do similar things.

Other browsers may still implement `protocol.LeakProtocolPart` directly if their concepts of leak checking are similar to Chrome's (i.e., can be expressed in terms of per-page named counters). But without standardization or other `LeakProtocolPart` implementors, we don't actually know yet if the `LeakProtocolPart` interface is useful for non-Chrome browsers.